### PR TITLE
Added SSD training example

### PR DESCRIPTION
**Pull request by awolant**
Signed-off-by: Albert Wolant <awolant@nvidia.com>

For now, DALICOCOIterator is just copied DALIGenericIterator beeing worked on in #282

### CODE REVIEW

File: docs/examples/pytorch/single_stage_detector/coco_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+        super(COCOPipeline, self).__init__(
+            batch_size=batch_size, device_id=device_id, num_threads=num_threads, seed = seed)
+
+        self.input = ops.COCOReader(
+            file_root = file_root,
+            annotations_file = annotations_file,
+            ratio=True, ltrb=True,
+            random_shuffle=True)
+        self.decode = ops.HostDecoder(device = "cpu", output_type = types.RGB)
```
[the remark this answers is not in the document]
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/coco_pipeline.py]: Done

**Comment by szalpal** [comment]
Are we ok with the example, that can't easily remove Dali? I remember, that it was one of QA's issues

**Comment by JanuszL** [comment]
> Are we ok with the example, that can't easily remove Dali? I remember, that it was one of QA's issues

We can point them to reference implementation https://github.com/mlperf/training/tree/master/single_stage_detector. So I think this will do.

File: docs/examples/pytorch/single_stage_detector/train.py [hunk 2]
```diff
@@ -185,7 +186,7 @@ def train300_mlperf_coco(args):
         dboxes,
         args.seed)
     train_pipe.build()
-    train_loader = DALICOCOIterator(train_pipe, ["images", "boxes", "labels"], 118287)
+    train_loader = DALIGenericIterator(train_pipe, ["images", "boxes", "labels"], 118287)
```
Comment (JanuszL) [inline docs/examples/pytorch/single_stage_detector/train.py]: Please try to use `pipe.epoch_size("Reader")` instead of hard-coded `118287`.
And in the COCO pipeline itself please named reader `inputs, bboxes, labels = self.input(name="Reader")`

Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/train.py]: Done

**Comment by awolant** [comment]
Builds 40749

File: qa/L1_ssd_training/test.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+pushd ../..
+
+cd docs/examples/pytorch/single_stage_detector/
+
+test_body() {
+    # workaround due to errors while pycocotools is int "pip_packages" above
+    pip install -I pycocotools
+
+    # test code
+    python train.py --data=/data/coco/coco-2017/coco2017/ --epochs=1
```
Comment (szalpal) [inline qa/L1_ssd_training/test.sh]: Is it possible to use distributed training here? It will be faster then

File: docs/examples/pytorch/single_stage_detector/train.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+                print("")
+                print("lr decay step #2")
+                for param_group in optim.param_groups:
+                    param_group['lr'] = current_lr
+                mlperf_log.ssd_print(key=mlperf_log.OPT_LR,
+                                     value=current_lr)
+
+            img = Variable(img, requires_grad=True)
+            ploc, plabel = ssd300(img)
+            trans_bbox = bbox.transpose(1,2).contiguous()
```
Comment (Kh4L) [inline docs/examples/pytorch/single_stage_detector/train.py]: I know I am nitpicking here, but you should move this to L234

File: docs/examples/pytorch/single_stage_detector/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+Run the run_and_time.sh script
+```
+cd reference/single_stage_detector/ssd
+source run_and_time.sh SEED TARGET
+```
+where SEED is the random seed for a run, TARGET is the quality target from Section 5 below.
+
+Docker Image
+```
+sudo nvidia-docker run -v /coco:/coco -t -i --rm --ipc=host mlperf/single_stage_detector ./run_and_time.sh SEED TARGET
```
[the remark this answers is not in the document]
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/README.md]: Right, I will change it in run_and_time.sh as this above is as MLPerf does it

File: docs/examples/pytorch/single_stage_detector/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
+# 1. Problem
```
[the remark this answers is not in the document]
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/README.md]: I can remove it, but it is copied as is from reference MLPerf

File: qa/L1_ssd_training/test.sh (hunk 3 above)
Comment (awolant) [reply qa/L1_ssd_training/test.sh]: Not really, this implementation is not prepared for distributed training. I was supposed to add support in next stage of work.

**Comment by JanuszL** [comment]
@Kh4L - this code is taken from https://github.com/mlperf/training/tree/master/single_stage_detector/ssd as is. I don't think we should rework it much beyond pluging DALI there (this includes code itself, readme and etc).

**Comment by awolant** [comment]
Things that were changed from reference MLPerf are: added coco_pipeline.py,  train.py (only parts with train loader). Also L1_ssd_training test was added.

File: docs/examples/pytorch/single_stage_detector/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (awolant) [reply docs/examples/pytorch/single_stage_detector/README.md]: Fixed in run_and_time.sh

**Comment by Kh4L** [comment]
@JanuszL  I reworked the RN-50 example to make it simpler to understand, but ok I am fine with leaving it as MLperf ref in this SSD example

**Comment by awolant** [comment]
Builds 40767
